### PR TITLE
Fix(slack): unanswered turn nudge

### LINKED ISSUE #3674: A question asked in Slack is answered in the web chat, so the person who asked never sees it
## Context

Slack is where most people reach an agent. A Slack turn only reaches the channel when the agent calls its reply tool. Plain assistant text is discarded. When the agent writes its answer as normal text, the work is done but the answer stays in the session.

## Problem

A user asks the agent something in Slack. The agent does the work and writes the answer, but the answer appears only in the agent's chat session on the web. Slack shows nothing. The user reads it as no answer, and asks again or gives up. This is silent — nothing marks the turn as failed.

## Expected Behavior

An answer to a Slack message appears in Slack, in the thread where the question was asked.

## Actual Behavior

The answer appears in the agent's session in the web UI. Slack stays empty.

## Steps to Reproduce

1. Connect an agent to a Slack channel.
2. Ask the agent a question in that channel.
3. Wait for the turn to finish.
4. Open the agent's session in the web UI — the answer is there. Slack has no reply.

### PR DESCRIPTION
**Pull request by jjeliga**
closes #3674

- Slack answers that the agent writes but forgets to send now get posted. It's asked once; if it stays quiet, that's the end.
- "No reply needed" counts as a proper ending, not a failure.
- It won't prod an agent that already answered, including when a follow-up message got answered alongside the first.
- Agents in ambient mode are never prodded unlless replying to a mention.
- The late reply reads as the answer, not an apology. If you were shown an error, it doesn't pretend otherwise.
- If the agent answers that thread again on its own before the reminder lands, the reminder stands down.

### CODE REVIEW

File: packages/api-server/src/modules/channels/infrastructure/slack.ts [hunk 1]
```diff
@@ -1535,73 +1651,9 @@ export function createSlackWorker(
...
-                channel: "slack",
-                agentId: instanceName,
-                actorSub: null,
-                outcome: answered ? "success" : "failure",
-                reason: "recovery-nudge",
-              });
-            }
-          },
-        });
+        watchTurn(sessionId, verdictRefs(), true);
```
Comment (dam-code-guardian) [inline packages/api-server/src/modules/channels/infrastructure/slack.ts]: 🟡 **Warning:** this call passes `sawFailure: true` unconditionally, but `postFailure` posts no failure copy when the turn already posted something (`slack.ts:1622`), and the interrupted verdict nudges exactly that turn — `deliveredBy(refs, "interrupted")` counts only `declined` and `handedOff`, never `posted`. So an abandoned turn with an interim post gets the notice that says the person "was already told the turn had gone wrong", and loses the "do not apologise for the delay" line: the late answer apologises for a failure nobody saw, which is the same mismatch this PR adds the parameter to prevent.
  **Fix:** derive `sawFailure` from whether the failure copy actually reached the thread — have `postFailure` report that it posted and pass that value here. A parameter that selects user-facing copy is read from the state it describes, never written as a literal at the call site.

File: packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts [hunk 2]
```diff
@@ -151,6 +166,10 @@ export function createTurnRecovery(deps: {
         deadline: Date.now() + RECOVERY_WINDOW_MS,
       };
       watches.set(key, state);
+      if (opts?.endedAs !== undefined) {
+        void finish(key, turn, opts.endedAs);
```
Comment (dam-code-guardian) [inline packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts]: 🟡 **Warning:** `finish` runs `remove(key)` before its first `await`, so a watch registered with `endedAs` never lives in `watches`: `dismiss` (`:177`), the `drop` of a re-registering `watch` and `stop` (`:181`) are all no-ops against an in-flight immediate recovery, and the `state` and `gen` stored one line above are never read. The addressed queue is keyed per sender and ambient traffic has its own queue, so a second turn on the same thread can already be waiting on `withSessionTurnLock` when the silent turn releases it; the nudge joins that queue only from the relay's `finally`, so it runs after that turn and then re-checks delivery over the silent turn's own refs alone. A later turn that answers therefore cannot dismiss the nudge and the person is answered twice — the invariant this module's own boundary comment states — and `stop()` cannot drain the nudge at shutdown.
  **Fix:** keep an immediate recovery addressable for its whole life: leave the generation-tagged state in `watches` until `finish` completes and retain its promise there, so `dismiss`, a re-registration and `stop` cancel or drain it exactly as they do a polled watch.

**Review by dam-code-guardian — COMMENTED** [review]
🛡️ **Code Guardian 🛡️** — ⚠️ Code Review @ `e57fcda`

## PR #3708: Fix(slack): unanswered turn nudge
**Author:** jjeliga | **Branch:** fix/3674-slack-unanswered-turn-nudge → main | **Changes:** +346 −78 (6 files)

### Summary
A cleanly-ended addressed Slack turn that posted nothing now goes to the same recovery watch as an abandoned one, registered with its ending already known, and the delivery verdict spans the refs steering mints mid-turn. One skill warning was refuted: `ensureReady` inside the turn lock matches `runSessionTurn`, which wakes the pod inside that same lock.
_5 suggestion(s) dropped under the 🟢 budget._

### Findings
- 🟡 **Warning:** the notice tells the agent the person saw failure copy that was never posted (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1656`)
- 🟡 **Warning:** an immediate recovery leaves the registry before `dismiss` can reach it (`packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts:170`)
- 🟢 **Suggestion:** the page keeps the `Last verified` date it had on `main` although this PR's newer commit edits it again, does not say that a reply to a steered message answers the turn, and still states one turn event per relay with a reason on failures only, which the nudge event now contradicts on the common path (`docs/architecture/channel-turns.md:3`, also `:55` and `:65`)
- ✅ **Looks good:** the snapshot in `verdictRefs()` is the correct choice, not a missed live read — `onTurnSettled` empties `steeredRefs` after the relay returns, so a live closure would judge delivery over an empty list, and `holdWatchedRefs` keeps the snapshot's refs resolvable for a late reply.

### Documentation Check (doc-drift)
- Both items are folded into the 🟢 above. Severity is graded here, not by the skill: on this repository a doc page that misses a stamp bump or a new clause is a suggestion, never a blocker.

### Issue Fit
- ✅ No findings that survive. The diff maps to #3674, and the ambient narrowing is declared in commit `bcdcd50` and held by a test, so it is a recorded descope. Its suggestion fell under the 🟢 budget.

### TypeScript Engineering Review
- 🟡 The uncancellable immediate recovery is reported in `### Findings` (`turn-recovery.ts:170`); the skill's separate fire-and-forget and dead-`watches.set` observations are the same defect and are merged into it.
- Refuted: `agents().ensureReady` inside `withSessionTurnLock`. `runSessionTurn` already wakes the pod inside that same lock, so the nudge follows the house pattern, and a message waiting on the thread pays that wake either way. Regraded 🟢, then dropped under the budget.
- Three further suggestions (tick-based assertions in the new cases, the options-bag tuple spread, no test pinning `resumeSessionId`) fell under the 🟢 budget.

### Verdict
COMMENT — both warnings sit in the new recovery path's own copy and lifecycle, and neither loses the answer the feature exists to deliver.

### For the human reviewer
The change gives one owner to both ways a turn can end with the answer unsent, which is the right call. It reaches that by letting the relay declare the ending itself and having the recovery judge it on the spot, instead of going through the watch's own life cycle. The cost of that shortcut is that the recovery is no longer something the platform can call off: the rule that a later answer to the same person cancels a pending nudge holds for one path and not for the other, and a shutdown cannot wait for it. Settle whether the shortcut keeps the cancel rule before the smaller copy fix matters.


---
_Review by [Code Guardian 🛡️](https://github.com/dam-agents/code-guardian) · automated code guardian_

**Comment by jjeliga** [comment]
Not fixing: `stop()` cancels an in-flight recovery but does not wait for it. Same on the polled path, predates this PR, wants its own issue.

File: packages/api-server/src/modules/channels/infrastructure/slack.ts (hunk 1 above)
Comment (jjeliga) [reply packages/api-server/src/modules/channels/infrastructure/slack.ts]: Correct, and it is the designed interrupted path rather than an edge case, so this would have fired in normal use.

Fixed in `d240bc7`. `postFailure` now reports whether it actually posted, and that value is what the call passes. Test added for the interrupted-with-a-partial-post case, confirmed to fail without the fix.

File: packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts (hunk 2 above)
Comment (jjeliga) [reply packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts]: You are right and my first reply was wrong. I claimed there was no window because the nudge takes the session lock synchronously. That misses the queue split you point at: an ambient thread reply resumes the same session key a mention there uses but runs on its own queue, so it can already be waiting on that lock when the silent turn releases it. The nudge joins behind it, that turn answers, and its dismissal finds nothing to cancel.

Fixed in `5b0894a2`, close to your suggestion. The watch stays in the registry while the recovery runs, and the recovery is handed a cancellation check that the nudge consults once it holds the session lock, so it stands down instead of answering someone who has been answered since. Test added, confirmed to fail without it.

Not doing the drain half: `stop()` now cancels an in-flight recovery but does not await it. That gap is the same on the polled path and predates this PR, so it belongs in its own issue.

Comment (dam-code-guardian) [reply packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts]: Thank you for the correction, and the queue split is the part I could not have shown from the code alone. Your fix is what the next review checks: the watch stays in the registry while the recovery runs, and the nudge reads the cancellation check after it takes the session lock.

I recorded the drain half as a PR-local decline for this pull request. `stop()` cancels an in-flight recovery and does not wait for it. I verified that the polled path has the same gap on `main`, so it is not a defect of this change and no later review of this pull request reports it. A separate issue is the correct place for it.

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **Code Guardian 🛡️** — ✅ Code Review @ `5b0894a`

### Summary
Both warnings of the last round are closed at their cause: the notice copy now follows whether the failure text actually reached the thread, and an immediate recovery stays in the registry so a dismissal or a shutdown reaches it. The page carries the new stamp and both missing clauses.
_7 suggestion(s) dropped under the 🟢 budget._

### Changes since last review
Previous HEAD: `e57fcda` (2026-09-16T07:04:23Z) — verdict COMMENT

- ✅ **Fixed:** notice claims failure copy the person never saw (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1656`)
- ✅ **Fixed:** immediate recovery leaves the registry before dismiss can reach it (`packages/api-server/src/modules/channels/infrastructure/turn-recovery.ts:170`)
- ✅ **Fixed:** stale page stamp and two unstated claims (`docs/architecture/channel-turns.md:3`)
- 🆕 **New:** failure copy gate reads the opening batch alone (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1625`)
- 🆕 **New:** page keys the notice variant on the path, the code on posted copy (`docs/architecture/channel-turns.md:67`)
- 🆕 **New:** nudge reuses the turn relayed event, so usage counts two turns (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1329`)

### Findings
- 🟢 **Suggestion:** `postFailure` asks `turnRefs.some(…)` over the opening batch, while the verdict and the log gate ask `deliveredBy(verdictRefs(), …)` over the batch plus the steered refs. An abandoned turn whose one reply landed on a steered message therefore gets platform failure copy on top of the answer the person already has, and the notice then correctly says they saw a failure, so the late answer apologises for it. This line is unchanged by the two new commits — the two ref sets parted at `e57fcda` — so it is a suggestion here; giving the failure copy the same `verdictRefs()` predicate closes it (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1625`)
- 🟢 **Suggestion:** the page keys the two notice variants on the path that reached them ("After an abandoned turn the notice says the opposite, because there the failure copy is already on screen"), while `d240bc7` keys them on `failurePosted` — the fact, not the path. The PR's own new test pins an abandoned turn that gets the no-failure variant, so the page states a rule this commit removed (`docs/architecture/channel-turns.md:67`)
- 🟢 **Suggestion:** the nudge announces itself as `ChannelTurnRelayed` with `reason: "recovery-nudge"`, and `record-usage.ts` calls `recorder.turn(…)` on every one of those events, so a repaired turn counts as two turns in usage metrics. The page now declares the two events, and no control flow reads them (`packages/api-server/src/modules/channels/infrastructure/slack.ts:1329`)

### Documentation Check (doc-drift)
- 🟢 The page-versus-code discriminator is reported in `### Findings` (`channel-turns.md:67`). Severity is graded here: on this repository a page that keeps a sentence a commit falsified is a suggestion, never a blocker.
- One further suggestion (the clean-finish nudge stated without its `ghostTurn` and missing-session guards) fell under the 🟢 budget.

### Issue Fit
- ✅ No findings that survive. The two suggestions fell under the 🟢 budget, and the ambient narrowing is the recorded descope of `bcdcd50`, held by its test.

### TypeScript Engineering Review
- 🟢 The stale delivery predicate and the overloaded turn event are reported in `### Findings` (`slack.ts:1625`, `slack.ts:1329`).
- Refuted: the optional `siblingRefs` narrowing the verdict. Both `relaySharedTurn` call sites are correct — the coalescer passes `() => steeredRefs`, and the hand-off relay carries one message, so no sibling can exist for it. Regraded 🟢, then dropped under the budget.
- Refuted: the layering item. `turn-recovery.ts` sits in `infrastructure/` on `main` too, so this PR moves nothing; regraded 🟢, then dropped.
- Two further suggestions (a discriminated union for `watch`, the copy export for the test) fell under the 🟢 budget.

### Verdict
APPROVE — the approval bar is empty: both warnings are closed at their cause, and the three suggestions are non-blocking.


---
_Review by [Code Guardian 🛡️](https://github.com/dam-agents/code-guardian) · automated code guardian_

**Review by pilartomas — APPROVED** [review]
